hydra-node/src/Hydra/PersistentQueue.hs

```diff
@@ -0,0 +1,148 @@
+module Hydra.PersistentQueue where
+
+import Hydra.Prelude
+
+import Cardano.Binary (decodeFull', serialize')
+import Control.Concurrent.Class.MonadSTM (
+  modifyTVar',
+  newTBQueueIO,
+  newTVarIO,
+  peekTBQueue,
+  readTBQueue,
+  writeTBQueue,
+ )
+import Control.Exception (IOException)
+import Data.Aeson (eitherDecode', encode)
+import Data.List qualified as List
+import System.Directory (createDirectoryIfMissing, doesPathExist, listDirectory, removeFile)
+import System.FilePath ((</>))
+import UnliftIO.IO.File (writeBinaryFileDurable)
+
+-- * Persistent queue
+
+data PersistentQueue m a = PersistentQueue
+  { queue :: TBQueue m (Natural, a)
+  , nextIx :: TVar m Natural
+  , directory :: FilePath
+  }
+
+-- | Create a new persistent queue at file path and given capacity.
+newPersistentQueue ::
+  (MonadSTM m, MonadIO m, FromCBOR a, MonadCatch m, MonadFail m) =>
+  FilePath ->
+  Natural ->
+  m (PersistentQueue m a)
+newPersistentQueue path capacity = do
+  queue <- newTBQueueIO capacity
+  highestId <-
+    try (loadExisting queue) >>= \case
+      Left (_ :: IOException) -> do
+        liftIO $ createDirectoryIfMissing True path
+        pure 0
+      Right highest -> pure highest
+  nextIx <- newTVarIO $ highestId + 1
+  pure PersistentQueue{queue, nextIx, directory = path}
+ where
+  loadExisting queue = do
+    paths <- liftIO $ listDirectory path
+    case sort $ mapMaybe readMaybe paths of
+      [] -> pure 0
+      idxs -> do
+        forM_ idxs $ \(idx :: Natural) -> do
+          bs <- readFileBS (path </> show idx)
+          case decodeFull' bs of
+            Left err ->
+              fail $ "Failed to decode item: " <> show err
+            Right item ->
+              atomically $ writeTBQueue queue (idx, item)
+        pure $ List.last idxs
+
+-- | Create a new persistent queue at file path and given capacity.
+newPersistentQueueJson ::
+  (MonadSTM m, MonadIO m, FromJSON a, MonadCatch m, MonadFail m) =>
+  FilePath ->
+  Natural ->
+  m (PersistentQueue m a)
+newPersistentQueueJson path capacity = do
+  queue <- newTBQueueIO capacity
+  highestId <-
+    try (loadExisting queue) >>= \case
+      Left (_ :: IOException) -> do
+        liftIO $ createDirectoryIfMissing True path
+        pure 0
+      Right highest -> pure highest
+  nextIx <- newTVarIO $ highestId + 1
+  pure PersistentQueue{queue, nextIx, directory = path}
+ where
+  loadExisting queue = do
+    paths <- liftIO $ listDirectory path
+    case sort $ mapMaybe readMaybe paths of
+      [] -> pure 0
+      idxs -> do
+        forM_ idxs $ \(idx :: Natural) -> do
+          bs <- readFileLBS (path </> show idx)
+          case eitherDecode' bs of
+            Left err ->
+              fail $ "Failed to decode item: " <> show err
+            Right item ->
+              atomically $ writeTBQueue queue (idx, item)
+        pure $ List.last idxs
```

Comment:
Must: not duplicate the logic just for JSON / CBOR encoding

If we want both, instead we should parameterize the `PersistentQueue` with it and just put `encode` and `decode` functions into the handle. Then, you would have:

```haskell
newPersistentQueue ::
  (MonadSTM m, MonadIO m, MonadCatch m, MonadFail m) =>
  -- | Encode
  (a -> m ()) ->
  -- | Decode
  (m (Either Text a)) ->
  FilePath ->
  Natural ->
  m (PersistentQueue m a)

newPersistentQueueJson ::
  (MonadSTM m, MonadIO m, FromJSON a, MonadCatch m, MonadFail m) =>
  FilePath ->
  Natural ->
  m (PersistentQueue m a)
newPersistentQueueJson = newPersistentQueue Aeson.encode Aeson.eitherDecode'
```